File: model-pipeline/gallery.py
"""Precompute the showcase gallery, THROUGH THE fp32 ONNX CHAIN.

The gallery must match what the live browser produces, so every number here comes
from running the segmented fp32 graphs in onnxruntime -- NOT from TransformerLens.
(validate.py already proved the fp32 chain matches TL to <1e-3; here we just want
byte-for-byte the same code path the web app runs.)

For each selected prompt we emit:
  - tokens        : display strings (decoded per-token text, real spaces/newlines)
  - token_ids     : the GPT-2 BPE ids
  - lens_top5     : logit lens -- unembed(resid_post[layer]) softmaxed, top-5 per
                    (layer, position): [layer][pos] -> [{"t": token_str, "p": prob}]
  - patterns_u8   : all attention patterns [12 layers, 12 heads, T, T] quantized to
                    uint8 (value/scale, scale = max/255) and base64'd, for compact
                    transport. shape + scale let the client reconstruct floats.

Usage:
    python gallery.py --artifacts D:/dev/interp-artifacts/onnx
"""

import argparse
import base64
import json
from pathlib import Path

import numpy as np
import onnxruntime as ort
from transformers import GPT2TokenizerFast

HERE = Path(__file__).parent
N_LAYERS = 12
GALLERY_IDS = ["pangram", "ioi", "induction", "factual"]


def load_sessions(variant_dir: Path):
    names = ["embed", "unembed"] + [f"block_{i:02d}" for i in range(N_LAYERS)]
    return {
        n: ort.InferenceSession(str(variant_dir / f"{n}.onnx"),
                                providers=["CPUExecutionProvider"])
        for n in names
    }


def run_chain(sessions, ids: np.ndarray):
    """Return per-layer resids (resid_post) and per-layer patterns."""
    resid = sessions["embed"].run(None, {"input_ids": ids})[0]
    resids, patterns = [], []
    for i in range(N_LAYERS):
        resid, pattern = sessions[f"block_{i:02d}"].run(None, {"resid": resid})
        resids.append(resid)
        patterns.append(pattern)
    return resids, patterns


def softmax_lastdim(x: np.ndarray) -> np.ndarray:
    x = x - x.max(axis=-1, keepdims=True)
    e = np.exp(x)
    return e / e.sum(axis=-1, keepdims=True)


def lens_top5(sessions, resids, tokenizer):
    """[layer][pos] -> list of {"t": token_str, "p": prob} for the top-5 tokens."""
    out = []
    for r in resids:  # r: [1, seq, 768]
        logits = sessions["unembed"].run(None, {"resid": r})[0][0]  # [seq, vocab]
        probs = softmax_lastdim(logits)
        top5 = np.argsort(-probs, axis=-1)[:, :5]  # [seq, 5]
        layer_rows = []
        for pos in range(top5.shape[0]):
            cells = [
                {"t": tokenizer.decode([int(tid)]), "p": round(float(probs[pos, tid]), 6)}
                for tid in top5[pos]
            ]
            layer_rows.append(cells)
        out.append(layer_rows)
    return out


def pack_patterns_u8(patterns):
    """Stack [12, 12, T, T] float patterns, quantize to uint8, base64-encode.

    scale = global_max / 255 (per prompt); u8 = round(value / scale). Reconstruct
    client-side as float = u8 * scale."""
    arr = np.stack([p[0] for p in patterns], axis=0)  # [12, 12, T, T]
    max_val = float(arr.max())
    scale = max_val / 255.0 if max_val > 0 else 1.0
    u8 = np.clip(np.round(arr / scale), 0, 255).astype(np.uint8)
    return {
        "shape": list(u8.shape),
        "scale": scale,
        "data": base64.b64encode(u8.tobytes()).decode("ascii"),
    }


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--artifacts", required=True,
                        help="parent dir containing fp32/ (gallery uses fp32 only)")
    args = parser.parse_args()
    fp32_dir = Path(args.artifacts) / "fp32"

    prompts = {p["id"]: p for p in
               json.loads((HERE / "prompts.json").read_text(encoding="utf-8"))["prompts"]}
    tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")
    sessions = load_sessions(fp32_dir)

    out_prompts = []
    for pid in GALLERY_IDS:
        p = prompts[pid]
        ids = tokenizer(p["text"], return_tensors="np")["input_ids"].astype(np.int64)
        token_ids = ids[0].tolist()
        tokens = [tokenizer.decode([tid]) for tid in token_ids]  # display strings

        resids, patterns = run_chain(sessions, ids)
        out_prompts.append({
            "id": pid,
            "text": p["text"],
            "tokens": tokens,
            "token_ids": token_ids,
            "lens_top5": lens_top5(sessions, resids, tokenizer),
            "patterns_u8": pack_patterns_u8(patterns),
        })
        print(f"  {pid}: {len(token_ids)} tokens")

    out_dir = HERE / "out"
    out_dir.mkdir(exist_ok=True)
    (out_dir / "gallery.json").write_text(
        json.dumps({"prompts": out_prompts}), encoding="utf-8")
    print(f"wrote {out_dir / 'gallery.json'} ({(out_dir / 'gallery.json').stat().st_size / 1e6:.2f} MB)")


if __name__ == "__main__":
    main()

File: web/src/lib/gallery.ts
/**
 * gallery.ts — optional precomputed-examples loader.
 *
 * If `/gallery.json` exists it lets the app show worked examples (attention +
 * logit lens) *instantly*, before/without the ~164MB+ model download. The file
 * is produced by the model pipeline and integrated later; a missing file must
 * never error — the app just falls back to the live-model path.
 *
 * Schema (frozen by the milestone brief):
 *   { prompts: [ {
 *       id, text,
 *       tokens: string[],            // display labels
 *       token_ids: number[],
 *       lens_top5: [layer][pos][{ t, p }],   // top-5 predicted token + prob
 *       patterns_u8: {               // quantized attention, per layer
 *         shape: [n_layers, n_heads, seq, seq],
 *         scale: number,             // real = u8 * scale
 *         data: string               // base64 of the u8 buffer
 *       }
 *   } ] }
 */

export interface GalleryLensCell {
  t: string
  p: number
}

export interface GalleryPatternsU8 {
  shape: [number, number, number, number] // [layers, heads, seq, seq]
  scale: number
  data: string // base64 uint8
}

export interface GalleryPrompt {
  id: string
  text: string
  tokens: string[]
  token_ids: number[]
  lens_top5: GalleryLensCell[][][] // [layer][pos][top5...]
  patterns_u8: GalleryPatternsU8
}

export interface Gallery {
  prompts: GalleryPrompt[]
}

/** Decode base64 (browser) into a Uint8Array. */
export function base64ToU8(b64: string): Uint8Array {
  const bin = atob(b64)
  const out = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i)
  return out
}

/**
 * Dequantize a gallery pattern block into per-layer Float32Array patterns in
 * the same [1, heads, seq, seq] layout the live runner produces, so the views
 * can consume either source identically.
 */
export function galleryPatternsToFloat(p: GalleryPatternsU8): Float32Array[] {
  const [layers, heads, s1, s2] = p.shape
  const u8 = base64ToU8(p.data)
  const perLayer = heads * s1 * s2
  const out: Float32Array[] = []
  for (let l = 0; l < layers; l++) {
    const f = new Float32Array(perLayer)
    const base = l * perLayer
    for (let i = 0; i < perLayer; i++) f[i] = u8[base + i] * p.scale
    out.push(f)
  }
  return out
}

/** Fetch /gallery.json. Returns null on any error or absence — never throws. */
export async function loadGallery(): Promise<Gallery | null> {
  try {
    const res = await fetch('/gallery.json', { cache: 'no-cache' })
    if (!res.ok) return null
    const json = (await res.json()) as Gallery
    if (!json || !Array.isArray(json.prompts)) return null
    return json
  } catch {
    return null
  }
}
